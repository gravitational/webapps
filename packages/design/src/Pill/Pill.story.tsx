import React from 'react';
import Pill from './Pill';

export default {
  title: 'Design/Pill',
};

export const PillOptions = () => {
  return (
    <>
      <Pill label="arch: x86_64" />
      <br />
      <br />
      <Pill label="hostname: ip-172-31-9-155.us-west-2.compute.internal" />
      <br />
      <br />
      <Pill label="arch: x86_64" dismissable />
      <br />
      <br />
      <Pill
        label="hostname: ip-172-31-9-155.us-west-2.compute.internal"
        dismissable
      />
    </>
  );
};
