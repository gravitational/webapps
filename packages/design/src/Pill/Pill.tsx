/**
 * Copyright 2022 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import styled from 'styled-components';

import { Cross } from '../Icon';

function Pill({ label, onDismiss }: Props) {
  const dismissable = !!onDismiss;
  return (
    <Wrapper dismissable={dismissable}>
      <Label>{label}</Label>
      <Dismiss
        dismissable={dismissable}
        onClick={(e: MouseEvent) => {
          e.stopPropagation();
          onDismiss(label);
        }}
      >
        <Cross />
      </Dismiss>
    </Wrapper>
  );
}

const Wrapper = styled.span`
  background: rgba(255, 255, 255, 0.05);
  border-radius: 35px;
  cursor: default;
  display: inline-block;
  padding: ${props => (props.dismissable ? '6px 6px 6px 14px;' : '6px 14px;')};
  white-space: nowrap;
`;

const Label = styled.span`
  display: inline;
`;

const Dismiss = styled.button`
  border-color: rgba(0, 0, 0, 0);
  background-color: rgba(0, 0, 0, 0);
  cursor: pointer;
  display: ${props => (props.dismissable ? 'inline-block' : 'none')};
`;

type Props = {
  label: string;
  dismissable?: boolean;
  onDismiss?: (labelName: string) => void;
};

export { Pill };
