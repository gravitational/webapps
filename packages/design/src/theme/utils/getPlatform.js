/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

export default function getPlatform() {
  if (typeof window !== 'undefined') {
    const userAgent = window.navigator.userAgent;

    return {
      isWin: userAgent.indexOf('Windows') >= 0,
      isMac: userAgent.indexOf('Macintosh') >= 0,
      isLinux: userAgent.indexOf('Linux') >= 0,
    };
  } else {
    const platform = process.platform;
    const isWin = platform === 'win32';
    const isMac = platform === 'darwin';

    return {
      isWin,
      isMac,
      isLinux: !isWin && !isMac,
    };
  }
}
