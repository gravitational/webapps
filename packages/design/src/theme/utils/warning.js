/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Copyright (c) 2014-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Similar to invariant but only logs a warning if the condition is not met.
 * This can be used to log issues in development environments in critical
 * paths. Removing the logging code for production environments will keep the
 * same logic and follow the same code paths.
 */

/*eslint no-empty: off */

const __DEV__ = process.env.NODE_ENV !== 'production';

var warning = function() {};

if (__DEV__) {
  var printWarning = function printWarning(format, args) {
    var len = arguments.length;
    args = new Array(len > 2 ? len - 2 : 0);
    for (var key = 2; key < len; key++) {
      args[key - 2] = arguments[key];
    }
    var argIndex = 0;
    var message =
      'Warning: ' +
      format.replace(/%s/g, function() {
        return args[argIndex++];
      });
    if (typeof console !== 'undefined') {
      window.console.error(message);
    }
    try {
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    } catch (x) {}
  };

  warning = function(condition, format, args) {
    var len = arguments.length;
    args = new Array(len > 2 ? len - 2 : 0);
    for (var key = 2; key < len; key++) {
      args[key - 2] = arguments[key];
    }
    if (format === undefined) {
      throw new Error(
        '`warning(condition, format, ...args)` requires a warning ' +
          'message argument'
      );
    }
    if (!condition) {
      printWarning.apply(null, [format].concat(args));
    }
  };
}

export default warning;
