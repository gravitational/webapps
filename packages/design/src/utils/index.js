/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import ReactDOM from 'react-dom';
import getOffset from 'dom-helpers/query/offset';
import getPosition from 'dom-helpers/query/position';
import getScrollTop from 'dom-helpers/query/scrollTop';

export function ownerDocument(node) {
  return (node && node.ownerDocument) || document;
}

export function ownerWindow(node, fallback = window) {
  const doc = ownerDocument(node);
  return doc.defaultView || doc.parentView || fallback;
}

export function getContainer(container, defaultContainer) {
  container = typeof container === 'function' ? container() : container;
  return ReactDOM.findDOMNode(container) || defaultContainer;
}

function getContainerDimensions(containerNode) {
  let width, height, scroll;

  if (containerNode.tagName === 'BODY') {
    width = window.innerWidth;
    height = window.innerHeight;

    scroll =
      getScrollTop(ownerDocument(containerNode).documentElement) ||
      getScrollTop(containerNode);
  } else {
    ({ width, height } = getOffset(containerNode));
    scroll = getScrollTop(containerNode);
  }

  return { width, height, scroll };
}

function getTopDelta(top, overlayHeight, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerScroll = containerDimensions.scroll;
  const containerHeight = containerDimensions.height;

  const topEdgeOffset = top - padding - containerScroll;
  const bottomEdgeOffset = top + padding - containerScroll + overlayHeight;

  if (topEdgeOffset < 0) {
    return -topEdgeOffset;
  } else if (bottomEdgeOffset > containerHeight) {
    return containerHeight - bottomEdgeOffset;
  } else {
    return 0;
  }
}

function getLeftDelta(left, overlayWidth, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerWidth = containerDimensions.width;

  const leftEdgeOffset = left - padding;
  const rightEdgeOffset = left + padding + overlayWidth;

  if (leftEdgeOffset < 0) {
    return -leftEdgeOffset;
  } else if (rightEdgeOffset > containerWidth) {
    return containerWidth - rightEdgeOffset;
  }

  return 0;
}

export function calculatePosition(
  placement,
  overlayNode,
  target,
  container,
  padding
) {
  const childOffset =
    container.tagName === 'BODY'
      ? getOffset(target)
      : getPosition(target, container);

  const { height: overlayHeight, width: overlayWidth } = getOffset(overlayNode);

  let positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop;

  if (placement === 'left' || placement === 'right') {
    positionTop = childOffset.top + (childOffset.height - overlayHeight) / 2;

    if (placement === 'left') {
      positionLeft = childOffset.left - overlayWidth;
    } else {
      positionLeft = childOffset.left + childOffset.width;
    }

    const topDelta = getTopDelta(
      positionTop,
      overlayHeight,
      container,
      padding
    );

    positionTop += topDelta;
    arrowOffsetTop = 50 * (1 - (2 * topDelta) / overlayHeight) + '%';
    arrowOffsetLeft = void 0;
  } else if (placement === 'top' || placement === 'bottom') {
    positionLeft = childOffset.left + (childOffset.width - overlayWidth) / 2;

    if (placement === 'top') {
      positionTop = childOffset.top - overlayHeight;
    } else {
      positionTop = childOffset.top + childOffset.height;
    }

    const leftDelta = getLeftDelta(
      positionLeft,
      overlayWidth,
      container,
      padding
    );

    positionLeft += leftDelta;
    arrowOffsetLeft = 50 * (1 - (2 * leftDelta) / overlayWidth) + '%';
    arrowOffsetTop = void 0;
  } else {
    throw new Error(
      `calcOverlayPosition(): No such placement of "${placement}" found.`
    );
  }

  return { positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop };
}
