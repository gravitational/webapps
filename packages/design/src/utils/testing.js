import React from 'react';
import {
  render as testingRender,
  screen,
  fireEvent,
} from '@testing-library/react';
import ThemeProvider from 'design/ThemeProvider';
import theme from 'design/theme';
import 'jest-styled-components';

export function render(component) {
  return testingRender(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export { screen, fireEvent, theme };
