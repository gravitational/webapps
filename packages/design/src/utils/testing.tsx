/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  render as testingRender,
  act,
  fireEvent,
  waitForElement,
} from '@testing-library/react';
import { screen, wait, prettyDOM, getByTestId } from '@testing-library/dom';
import { MemoryRouter as Router } from 'react-router-dom';
import ThemeProvider from 'design/ThemeProvider';
import theme from 'design/theme';
import '@testing-library/jest-dom';
import 'jest-styled-components';

function Providers({ children }: { children: React.ReactElement }) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

function render(ui: React.ReactElement<any>, options?: RenderOptions) {
  return testingRender(ui, { wrapper: Providers, ...options });
}

screen.debug = () => {
  window.console.log(prettyDOM());
};

type RenderOptions = {
  wrapper: React.FC;
  container: HTMLElement;
};

export {
  act,
  screen,
  wait,
  fireEvent,
  theme,
  render,
  prettyDOM,
  waitForElement,
  getByTestId,
  Router,
};
