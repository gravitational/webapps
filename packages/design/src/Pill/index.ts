import Pill from "./Pill";

export default Pill