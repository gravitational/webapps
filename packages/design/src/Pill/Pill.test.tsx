/**
 * Copyright 2022 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import { render, fireEvent } from 'design/utils/testing';

import { Pill } from './Pill';

describe('design/Pill', () => {
  it('renders the label without dismissable', () => {
    const { container } = render(<Pill label="arch: x86_64" />);
    expect(container).toHaveTextContent('arch: x86_64');
    expect(container.getElementsByTagName('button')).toMatchSnapshot();
  });

  it('render the label with dismissable', () => {
    const { container } = render(
      <Pill label="arch: x86_64" onDismiss={() => {}} />
    );
    expect(container).toHaveTextContent('arch: x86_64');
    expect(container.getElementsByTagName('button')).toMatchSnapshot();
  });

  it('dismissing pill calls onDismiss', () => {
    const cb = jest.fn();
    const { container } = render(<Pill label="arch: x86_64" onDismiss={cb} />);
    fireEvent.click(container.querySelector('button'));
    expect(cb.mock.calls).toHaveLength(1);
    expect(cb.mock.calls).toEqual([['arch: x86_64']]);
  });
});
