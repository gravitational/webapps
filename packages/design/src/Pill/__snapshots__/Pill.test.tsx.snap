// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design/Pill render the label with dismissable 1`] = `
HTMLCollection [
  .c1 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c0 {
  border-color: rgba(0,0,0,0);
  background-color: rgba(0,0,0,0);
  cursor: pointer;
  display: inline-block;
}

<button
    class="c0"
  >
    <span
      class="c1 icon icon-cross "
      color="light"
    />
  </button>,
]
`;

exports[`design/Pill renders the label without dismissable 1`] = `
HTMLCollection [
  .c1 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c0 {
  border-color: rgba(0,0,0,0);
  background-color: rgba(0,0,0,0);
  cursor: pointer;
  display: none;
}

<button
    class="c0"
  >
    <span
      class="c1 icon icon-cross "
      color="light"
    />
  </button>,
]
`;
